mod/01_filter/set/abstract/filter.rb

```diff
@@ -0,0 +1,232 @@
+  end
+
+  def page_link_params
+    []
+  end
+
+  view :no_search_results do |_args|
+    %(
+      <div class="search-no-results">
+        No result
+      </div>
+    )
+  end
+
+  view :filter_form do |args|
+    formgroups = args[:formgroups] || [:name_formgroup]
+    html = formgroups.map do |fg|
+      optional_render(fg, args)
+    end
+    content = output(html)
+    action = card.left.name
+    %( <form action="/#{action}" method="GET">#{content}</form>)
+  end
+
+  def page_link text, page, _current=false, options={}
+    @paging_path_args[:offset] = page * @paging_limit
+    filter_args = {}
+    page_link_params.each do |key|
+      filter_args[key] = params[key] if params[key].present?
+    end
+    options[:class] = "card-paging-link slotter"
+    options[:remote] = true
+    link_to raw(text), path(@paging_path_args.merge(filter_args)), options
+  end
+
+  def text_filter type_name, args
+    formgroup args[:title] || type_name.capitalize,
+              text_field_tag(type_name, params[type_name],
+                             class: "form-control"),
+              class: " filter-input"
+  end
+
+  def type_options type_codename
+    type_card = Card[type_codename]
+    Card.search type_id: type_card.id, return: :name, sort: "name"
+  end
+
+  def select_filter type_codename, label=nil
+    label ||= type_codename.to_s.capitalize
+    options = type_options type_codename
+    options.unshift(["--", ""])
+    simple_select_filter type_codename.to_s, options, Env.params[type_codename],
+                         label
+  end
+
+  def simple_select_filter type_name, options, default, label=nil
+    options = options_for_select(options, default)
+    label ||= type_name.capitalize
+    formgroup label, select_tag(type_name, options, class: "pointer-select"),
+              class: "filter-input "
+  end
+
+  def simple_multiselect_filter type_name, options, default, label=nil
+    options = options_for_select(options, default)
+    label ||= type_name.capitalize
+    multiselect_tag = select_tag(type_name, options,
+                                 multiple: true,
+                                 class: "pointer-multiselect")
+    formgroup(label, multiselect_tag, class: "filter-input #{type_name}")
+  end
+
+  def multiselect_filter type_codename, label=nil
+    options = type_options type_codename
+    label ||= type_codename.to_s
+    simple_multiselect_filter type_codename.to_s, options,
+                              Env.params[type_codename], label
+  end
+
+  view :name_formgroup do |args|
+    name = args[:name] || "Name"
+    text_filter name, title: "Name"
+  end
+
+  view :project_formgroup do
+    select_filter :project
+  end
+
+  view :year_formgroup do
+    select_filter :year
+  end
+
+  view :topic_formgroup do
+    select_filter :wikirate_topic, "Topic"
+  end
+
+  view :metric_formgroup do
+    select_filter :metric
+  end
+
+  view :company_formgroup do
+    select_filter :wikirate_company, "Company"
+  end
```

Comment:
I thought these can be overridden easily later if necessary. We can only define the views we need in method #default_filter_form_args instead of calling the rendering everywhere. I can change it if you have a better idea.